# Creación de un modelo basado en Fine-Tuning de un modelo de terceros


En este notebook se realizará un procedimiento de Fine-Tunning sobre un modelo de terceros con el fin de batir los resultados de los anteriores modelos construidos


In [1]:
import sys

sys.path.append("../")  # Add the parent directory to the Python path

In [2]:
from src.data.data_loader import data_loader, data_loader_test
from src.models.train import ModelTrainer, ModelConfiguration
from src.models.evaluator import ModelEvaluator

## Carga de datos de entrenamiento


In [3]:
train_data, val_data = data_loader("../data/train", (150, 150), batch_size=128)

Found 12777 files belonging to 6 classes.
Using 10222 files for training.
Using 2555 files for validation.


## Carga del modelo preentrenado desde el catálogo de Keras


Se ha incluido dentro de la clase `ModelTrainer` la posibilidad de elegir entrenar un modelo basado en `Xception`.

#### Xception

La decisión de utilizar `Xception` frente a otros modelos de visión artificial viene motivada por los siguientes puntos:

- **Bajo número de parámetros**: A pesar de su alto rendimiento, `Xception` tiene un número relativamente bajo de parámetros en comparación con otras arquitecturas profundas. Esto hace que el modelo sea eficiente en cuanto a memoria y recursos computacionales, lo que es beneficioso para el desarrollo de esta práctica, ya que se cuenta con una modesta RTX 3060.

- **Características no probadas hasta ahora**: Xception` incluye en su arquitectura características innovadoras, como las convoluciones separables en profundidad y los módulos de afinidad lineal. Estas características no se habían probado en los modelos simple y aumentado, por lo que es interesante que a la hora de eligir un modelo externo se apueste por uno con funciones diferentes.


In [4]:
# El número de capas de Xception a entrenar es de 5, ya que no se quiere hacer fine tunning de demasiadas capas.
# Utilizando el equipo portátil del que se dispone no es recomendable incrementar este número

# El total de parámetros a entrenar del modelo es de 9.729.094, lo que es asumible
N_LAYERS_TO_TRAIN = 5

In [5]:
custom_xception_model = ModelTrainer(
    "xception_finetuned",
    ModelConfiguration.XCEPTION_CUSTOM,
    6,
)

In [6]:
custom_xception_model.train(
    train_data, val_data, epochs=25, n_layers_to_train=N_LAYERS_TO_TRAIN
)

COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: tensorboard, tensorflow, keras.
COMET WARNING: As you are running in a Jupyter environment, you will need to call `experiment.end()` when finished to ensure all metrics and code are logged before exiting.
COMET INFO: Couldn't find a Git repository in 'c:\\Users\\pablo\\Documents\\Master\\Modulo_8\\notebooks' nor in any parent directory. Set `COMET_GIT_DIRECTORY` if your Git Repository is elsewhere.
COMET INFO: Experiment is live on comet.com https://www.comet.com/pablotorresr/modulo8/2d576540d11f47f69981532e960c8160



Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_1 (InputLayer)           [(None, 150, 150, 3  0           []                               
                                )]                                                                
                                                                                                  
 block1_conv1 (Conv2D)          (None, 74, 74, 32)   864         ['input_1[0][0]']                
                                                                                                  
 block1_conv1_bn (BatchNormaliz  (None, 74, 74, 32)  128         ['block1_conv1[0][0]']           
 ation)                                                                                           
                                                                                              

INFO:tensorflow:Assets written to: bestvalue\assets


INFO:tensorflow:Assets written to: bestvalue\assets


80/80 [==============================] - 219s 3s/step - loss: 1.0970 - accuracy: 0.5861 - val_loss: 1.2003 - val_accuracy: 0.5750
Epoch 2/25
80/80 [==============================] - ETA: 0s - loss: 0.7376 - accuracy: 0.7170

INFO:tensorflow:Assets written to: bestvalue\assets


INFO:tensorflow:Assets written to: bestvalue\assets


80/80 [==============================] - 219s 3s/step - loss: 0.7376 - accuracy: 0.7170 - val_loss: 0.8167 - val_accuracy: 0.7037
Epoch 3/25
80/80 [==============================] - 194s 2s/step - loss: 0.6104 - accuracy: 0.7742 - val_loss: 0.8271 - val_accuracy: 0.7108
Epoch 4/25
80/80 [==============================] - 197s 2s/step - loss: 0.5128 - accuracy: 0.8060 - val_loss: 0.8262 - val_accuracy: 0.7108
Epoch 5/25
80/80 [==============================] - ETA: 0s - loss: 0.4500 - accuracy: 0.8346

INFO:tensorflow:Assets written to: bestvalue\assets


INFO:tensorflow:Assets written to: bestvalue\assets


80/80 [==============================] - 228s 3s/step - loss: 0.4500 - accuracy: 0.8346 - val_loss: 0.8148 - val_accuracy: 0.7198
Epoch 6/25
80/80 [==============================] - ETA: 0s - loss: 0.3577 - accuracy: 0.8702

INFO:tensorflow:Assets written to: bestvalue\assets


INFO:tensorflow:Assets written to: bestvalue\assets


80/80 [==============================] - 243s 3s/step - loss: 0.3577 - accuracy: 0.8702 - val_loss: 0.7813 - val_accuracy: 0.7378
Epoch 7/25
80/80 [==============================] - ETA: 0s - loss: 0.3171 - accuracy: 0.8826

INFO:tensorflow:Assets written to: bestvalue\assets


INFO:tensorflow:Assets written to: bestvalue\assets


80/80 [==============================] - 237s 3s/step - loss: 0.3171 - accuracy: 0.8826 - val_loss: 0.7731 - val_accuracy: 0.7315
Epoch 8/25
80/80 [==============================] - 206s 3s/step - loss: 0.2417 - accuracy: 0.9119 - val_loss: 0.9160 - val_accuracy: 0.7413
Epoch 9/25
80/80 [==============================] - 208s 3s/step - loss: 0.2002 - accuracy: 0.9253 - val_loss: 0.9948 - val_accuracy: 0.7100
Epoch 10/25
80/80 [==============================] - 209s 3s/step - loss: 0.1969 - accuracy: 0.9283 - val_loss: 1.0295 - val_accuracy: 0.7350
Epoch 11/25
80/80 [==============================] - 209s 3s/step - loss: 0.1415 - accuracy: 0.9489 - val_loss: 1.0040 - val_accuracy: 0.7550
Epoch 12/25
80/80 [==============================] - 210s 3s/step - loss: 0.1455 - accuracy: 0.9503 - val_loss: 1.1397 - val_accuracy: 0.7295
Epoch 13/25
80/80 [==============================] - 210s 3s/step - loss: 0.1066 - accuracy: 0.9626 - val_loss: 1.0553 - val_accuracy: 0.7511
Epoch 14/25
80/80 [=

COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     url                   : https://www.comet.com/pablotorresr/modulo8/2d576540d11f47f69981532e960c8160
COMET INFO:   Uploads:
COMET INFO:     conda-environment-definition : 1
COMET INFO:     conda-info                   : 1
COMET INFO:     conda-specification          : 1
COMET INFO:     environment details          : 1
COMET INFO:     filename                     : 1
COMET INFO:     installed packages           : 1
COMET INFO:     notebook                     : 1
COMET INFO:     source_code                  : 1
COMET INFO: 
COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: tensorboard, tensorflow, keras.
COMET INFO: Please wait

INFO:tensorflow:Assets written to: ../models/xception_finetuned\assets


INFO:tensorflow:Assets written to: ../models/xception_finetuned\assets


## Test del modelo Xception con Fine-Tuning


In [10]:
SAVED_MODEL_ROUTE = "../models/xception_finetuned"
test_data = data_loader_test("../data/test", (150, 150))

Found 4257 files belonging to 6 classes.


In [11]:
model_evaluator = ModelEvaluator(SAVED_MODEL_ROUTE)

In [12]:
model_evaluator.evaluate(test_data)

COMET WARNING: To get all data logged automatically, import comet_ml before the following modules: tensorboard, tensorflow, keras.
COMET WARNING: As you are running in a Jupyter environment, you will need to call `experiment.end()` when finished to ensure all metrics and code are logged before exiting.
COMET INFO: Couldn't find a Git repository in 'c:\\Users\\pablo\\Documents\\Master\\Modulo_8\\notebooks' nor in any parent directory. Set `COMET_GIT_DIRECTORY` if your Git Repository is elsewhere.
COMET INFO: Experiment is live on comet.com https://www.comet.com/pablotorresr/modulo8/9b9480d91d76479190fab4f68c3fe913



134/134 [==============================] - 73s 535ms/step - loss: 1.1878 - accuracy: 0.7496


COMET WARNING: Cannot safely convert [1.1878060102462769, 0.7495889067649841] object to a scalar value, using its string representation for logging. Resulting string might be invalid
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO: Comet.ml Experiment Summary
COMET INFO: ---------------------------------------------------------------------------------------
COMET INFO:   Data:
COMET INFO:     display_summary_level : 1
COMET INFO:     url                   : https://www.comet.com/pablotorresr/modulo8/9b9480d91d76479190fab4f68c3fe913
COMET INFO:   Metrics:
COMET INFO:     test_metrics : [1.1878060102462769, 0.7495889067649841]
COMET INFO:   Uploads:
COMET INFO:     conda-environment-definition : 1
COMET INFO:     conda-info                   : 1
COMET INFO:     conda-specification          : 1
COMET INFO:     environment details          : 1
COMET INFO:     filename                     : 1
COMET INFO:     installed packages  

#### Análisis de resultados

El resultado final de evaluar el modelo con Fine Tuning sobre el set de test de un accuracy del `0.7496`. </br>
Esta métrica es inferior a la obtenida en nuestro modelo optimizando los hiperparámetros. Para mejorar este bajo rendimiento se propone lo siguiente:

- Añadir más capas densas en el tramo final, para poder que la información proveniente de Xception se acople mejor

- Descongelar más capas de Xception

Ambas soluciones implican un gran coste computacional, por lo que es mejor antes probar otras alternativas, como entrenar el modelo con un set de imágenes aumentadas.

#### Desventajas adicionales del modelo

Debido a que el modelo cuenta con muchas más capas y neuronas que los anteriores, se ha observado que el tiempo de inferencia es también bastante superior. `40ms vs 400ms` por inferencia
